core/unix/rseq_linux.c

```diff
@@ -150,6 +187,79 @@ rseq_get_region_info(app_pc pc, app_pc *start OUT, app_pc *end OUT, app_pc *hand
     return true;
 }
 
+int
+rseq_get_tls_ptr_offset(void)
+{
+    /* This read is assumed to be atomic. */
+    ASSERT(rseq_tls_offset != 0);
+    return rseq_tls_offset + offsetof(struct rseq, rseq_cs);
+}
+
+static void
+rseq_clear_tls_ptr(dcontext_t *dcontext)
+{
+    byte *base = get_segment_base(LIB_SEG_TLS);
+    struct rseq *app_rseq = (struct rseq *)(base + rseq_tls_offset);
+    /* We're directly writing this in the cache, so we do not bother with safe_read
+     * or safe_write here either.  We already cannot handle rseq adversarial cases.
+     */
+    if (is_dynamo_address((byte *)(ptr_uint_t)app_rseq->rseq_cs.ptr64))
+        app_rseq->rseq_cs.ptr64 = 0;
+}
+
+int
+rseq_get_signature(void)
+{
+    /* This read is assumed to be atomic. */
```

Comment:
I would have to change it to an output variable instead of return value to use an atomic operation.  But I am agreeing that it doesn't matter here: if the signature has not been initialized and its initialization is racing with this query, it's just as bad to read a 0 (uninit) as to read a half-word of init and a half-word of uninit.  I will remove the comment and maybe replace w/ this logic.